core/src/main/java/org/testcontainers/dockerclient/transport/okhttp/OkHttpDockerCmdExecFactory.java

```diff
@@ -0,0 +1,152 @@
+package org.testcontainers.dockerclient.transport.okhttp;
+
+import com.github.dockerjava.api.command.PingCmd;
+import com.github.dockerjava.core.AbstractDockerCmdExecFactory;
+import com.github.dockerjava.core.DockerClientConfig;
+import com.github.dockerjava.core.SSLConfig;
+import com.github.dockerjava.core.WebTarget;
+import com.github.dockerjava.core.exec.PingCmdExec;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.MultimapBuilder;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import okhttp3.ConnectionPool;
+import okhttp3.Dns;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.internal.Internal;
+import org.apache.commons.io.IOUtils;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.X509TrustManager;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.URI;
+import java.security.cert.X509Certificate;
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+public class OkHttpDockerCmdExecFactory extends AbstractDockerCmdExecFactory {
+
+    private static final String SOCKET_SUFFIX = ".socket";
+
+    private OkHttpClient okHttpClient;
+
+    private HttpUrl baseUrl;
+
+    @Override
+    @SneakyThrows
+    public void init(DockerClientConfig dockerClientConfig) {
+        super.init(dockerClientConfig);
+
+        OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder()
+            .readTimeout(0, TimeUnit.SECONDS)
+            .retryOnConnectionFailure(true);
+
+        URI dockerHost = dockerClientConfig.getDockerHost();
+        switch (dockerHost.getScheme()) {
+            case "unix":
+            case "npipe":
+                String socketPath = dockerHost.getPath();
+
+                if ("unix".equals(dockerHost.getScheme())) {
+                    clientBuilder
+                        .socketFactory(new UnixSocketFactory(socketPath));
+                } else {
+                    clientBuilder
+                        .socketFactory(new NamedPipeSocketFactory(socketPath));
+                }
+
+                clientBuilder
+                    // Disable pooling
+                    .connectionPool(new ConnectionPool(0, 1, TimeUnit.SECONDS))
+                    .dns(hostname -> {
+                        if (hostname.endsWith(SOCKET_SUFFIX)) {
+                            return Collections.singletonList(InetAddress.getByAddress(hostname, new byte[]{0, 0, 0, 0}));
+                        } else {
+                            return Dns.SYSTEM.lookup(hostname);
+                        }
+                    });
+        }
+
+        SSLConfig sslConfig = dockerClientConfig.getSSLConfig();
+        if (sslConfig != null) {
+            SSLContext sslContext = sslConfig.getSSLContext();
+            if (sslContext != null) {
+                clientBuilder
+                    .sslSocketFactory(sslContext.getSocketFactory(), new TrustAllX509TrustManager());
```

Comment:
Not relevant for this PR, but I'm curious: why is the TrustAllX509TrustManager necessary?

Reply:
because it is mutual TLS and we have to trust the certs from Docker as well :)